api/bindings/v1alpha1/bindingconfiguration_types.go

```diff
@@ -0,0 +1,143 @@
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+package v1alpha1
+
+import (
+	"github.com/ngrok/ngrok-api-go/v5"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// NOTE: Run "make manifests" to regenerate code after modifying this file:w
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// common ngrok API/Dashboard fields
+type ngrokAPICommon struct {
+	// Description is a human-readable description of the object in the ngrok API/Dashboard
+	// +kubebuilder:validation:Required
+	// +kubebuilder:default:=`Created by ngrok-operator`
+	// +kubebuilder:validation:MaxLength=4096
+	Description string `json:"description,omitempty"`
+	// Metadata is a sJSON encoded tring of arbitrary data associated with the object in the ngrok API/Dashboard
+	// +kubebuilder:validation:Required
+	// +kubebuilder:default:=`{"owned-by":"ngrok-operator"}`
+	// +kubebuilder:validation:MaxLength=4096
+	Metadata string `json:"metadata,omitempty"`
+}
+
+// BindingConfigurationSpec defines the desired state of BindingConfiguration
+type BindingConfigurationSpec struct {
+	ngrokAPICommon `json:",inline"`
+
+	// Name is the name of the k8s-binding for the account to bind to this configuration and the ngrok API
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Pattern=`^k8s[/][a-zA-Z0-9-]{1,63}$`
+	Name string `json:"name"`
+
+	// AllowedURLs is a list of URI patterns ([scheme://]<service-name>.<namespace-name>) thet determine which EndpointBindings are allowed to be created by the operator
+	// TODO(hkatz) We are only implementing `*` for now
+	// Support more patterns in the future, see product spec
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:items:Pattern=`^[*]$`
+	AllowedURLs []string `json:"allowed_urls"`
+
+	// CertsSecretName is the name of the k8s secret that contains the TLS certificate, CSR, private/public keys to use for the ngrok API
+	// TODO(hkatz) Create controller to manage this Secret lifecycle
+	// +kubebuilder:validation:Required
+	// +kubebuilder:default="global-binding-configuration-certs"
+	CertsSecretName string `json:"certs_secret_name"`
```

Comment:
Instead of storing the `PublicKey` here directly I chose to use a Secret reference. Is this how we do the references? Or is there another pattern I should use?

Reply:
We currently have this pattern if you are extracting a specific key out of a k8s secret: https://github.com/ngrok/ngrok-operator/blob/1e2367de30c22a043fac83b7c946eb24925d4423/api/ingress/v1alpha1/ngrok_common.go#L217

It sounds like this reference is for creating a new k8s secret with `type=kubernetes.io/tls`. Is that correct?

Reply:
I'm not sure if it's exactly `type=kubernetes.io/tls` because we want to store the CSR too, but I do think I can use this `ClientSecret *SecretKeyRef` pattern for this field. Thanks!